### DISCUSSION TITLE
Issue with tracked url redirect to final url

### DISCUSSION THREAD

**Discussion by anebi** [postalserver/postal#2326] [Help with using Postal]
Hello,

I noticed a problem related to click tracking.

When I activate it, it changes the url correctly and when i click to the url it tracks correctly. This part is working perfectly.

The problem happens when I click on the url and the Postal service should redirect to the final url. Something is wrong here because many times the redirects are not working properly. It could be because it doesn't redirect with all query params or something similar.

Best regards,
Ali Nebi

**Comment by anebi** [comment]
Could be related to this ticket: https://github.com/postalserver/postal/issues/907

**Comment by willpower232** [comment]
You haven't said what does happen if the redirect doesn't happen but I would guess that the `X-Postal-Track-Host` header is missing from your caddy or nginx configuration

https://docs.postalserver.io/features/click-and-open-tracking#configuring-your-web-server

**Comment by anebi** [comment]
Hello,

Thanks for your reply. 

Yes, I do set each vhost with that header and the tracking is working. When i click to the shorten URL the redirect happens, just the final url where I am redirected is malformed or with missing query params, which makes the redirect bad.

I think that It related to the other issue I posted, related to & and '&amp;' but I need to retest to confirm this.

Right now I keep click tracking disabled because of this issue.

**Reply by willpower232** [reply]
Interesting, I think my fix for that did eventually get merged https://github.com/postalserver/postal/pull/296

I presume you're definitely using Postal V2 via docker so you should have this fix and it would be good to see what your testing reveals.

**Reply by anebi** [reply]
I have this version installed:

root@psrv:~# postal upgrade
warning: redirecting to https://github.com/postalserver/install/
From https://postalserver.io/start/install
 * branch            main       -> FETCH_HEAD
Already up to date.
No version specified, using latest available version...
Upgrading to 2.1.2
Pulling web      ... done
Pulling smtp     ... done
Pulling worker   ... done
Pulling cron     ... done
Pulling requeuer ... done
Creating postal_runner_run ... done

Should I change to use the master/main branch instead?

**Reply by willpower232** [reply]
you should be fine, can you verify exactly which URLs don't work?

**Reply by anebi** [reply]
Yes, let me run some tests and will get back to you with more specific info. I will try to reproduce the problem again :)

**Reply by anebi** [reply]
I was able to reproduce the problem. Here is an example:

If you compare both urls you will notice that & and = are replaced with their encoded versions.

```
'&' => &amp;
'=' => &#61;
```

```
working url
------
https://mydomain.tld/auth/realms/master/login-actions/action-token?key=eyJhbGciOiJIUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICI1NTg5YjdmYy1mZWEwLTRlNDItYjBmMy03YjAzYWZkYmMxNzgifQ.eyJqdGkiOiJjZGQ4MzZjZi0wZDk3LTRlZTMtYmEyZC04YjdmYTEyNjNlZjkiLCJleHAiOjE2NzQxMzg4MDYsIm5iZiI6MCwiaWF0IjoxNjc0MTM4NTA2LCJpc3MiOiJodHRwczovL2tjLmlndWFuYWl0LmNvbS9hdXRoL3JlYWxtcy9tYXN0ZXIiLCJhdWQiOiJodHRwczovL2tjLmlndWFuYWl0LmNvbS9hdXRoL3JlYWxtcy9tYXN0ZXIiLCJzdWIiOiJjZTdhN2FhOC1hNTc4LTQ3MjItYjVkZC0yM2I3NTFiMzgwYmQiLCJ0eXAiOiJ2ZXJpZnktZW1haWwiLCJhenAiOiJzZWN1cml0eS1hZG1pbi1jb25zb2xlIiwibm9uY2UiOiJjZGQ4MzZjZi0wZDk3LTRlZTMtYmEyZC04YjdmYTEyNjNlZjkiLCJlbWwiOiJhbmViaUBmaW5pa28uY29tIiwiYXNpZCI6ImNmMDg0Zjg5LTIxODYtNDhkYi05ZGJlLTk0YThkZTBmZGE3My5rSzV4X3BaaW4wWS5mYmM0NDM5YS04ZTlmLTQ1YmEtYTU2ZC01NTQ5ZjFlYmRlZWIiLCJhc2lkIjoiY2YwODRmODktMjE4Ni00OGRiLTlkYmUtOTRhOGRlMGZkYTczLmtLNXhfcFppbjBZLmZiYzQ0MzlhLThlOWYtNDViYS1hNTZkLTU1NDlmMWViZGVlYiJ9.0Pbnlg1TIgZsdYbfwdoW5gCRKp1OI7QeEnhYAWLoEdY&client_id=security-admin-console&tab_id=kK5x_pZin0Y


bad url (redirected to this url after clicking the shorten url)
------
https://mydomain.tld/auth/realms/master/login-actions/action-token?key&#61;eyJhbGciOiJIUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICI1NTg5YjdmYy1mZWEwLTRlNDItYjBmMy03YjAzYWZkYmMxNzgifQ.eyJqdGkiOiJlNWQ0MGFjOS1lYzI4LTQ4NDYtOWM1YS0wZDNhNDUyYjM3NTgiLCJleHAiOjE2NzQxMzkwNjksIm5iZiI6MCwiaWF0IjoxNjc0MTM4NzY5LCJpc3MiOiJodHRwczovL2tjLmlndWFuYWl0LmNvbS9hdXRoL3JlYWxtcy9tYXN0ZXIiLCJhdWQiOiJodHRwczovL2tjLmlndWFuYWl0LmNvbS9hdXRoL3JlYWxtcy9tYXN0ZXIiLCJzdWIiOiIwYjQyOGE3ZC1hMWQ4LTQ1MjEtYWJhYS1lY2E0MDk4Zjg3ZTEiLCJ0eXAiOiJ2ZXJpZnktZW1haWwiLCJhenAiOiJzZWN1cml0eS1hZG1pbi1jb25zb2xlIiwibm9uY2UiOiJlNWQ0MGFjOS1lYzI4LTQ4NDYtOWM1YS0wZDNhNDUyYjM3NTgiLCJlbWwiOiJhbmViaUBmaW5pa28uY29tIiwiYXNpZCI6IjE3MzM1NTJkLWRjOTctNDhkNC1iZjRhLTY4MDk4NTI4MjgwNC5JalZZOXc4UGkxRS5mYmM0NDM5YS04ZTlmLTQ1YmEtYTU2ZC01NTQ5ZjFlYmRlZWIiLCJhc2lkIjoiMTczMzU1MmQtZGM5Ny00OGQ0LWJmNGEtNjgwOTg1MjgyODA0LklqVlk5dzhQaTFFLmZiYzQ0MzlhLThlOWYtNDViYS1hNTZkLTU1NDlmMWViZGVlYiJ9.Wvoz4rDyy24rlMzksTuuniJm5ckefm4H0i9IODATUsc&amp;client_id&#61;security-admin-console&amp;tab_id&#61;IjVY9w8Pi1E
```

**Reply by anebi** [reply]
Hello,

I don't see any reply after my latest comment here. Is there any progress with this problem?

Regards,
Ali Nebi

**Reply by anebi** [reply]
Hello, @willpower232 do you have any news about this issue?

Thank you in advance!

Best regards,
Ali Nebi

**Comment by anebi** [comment]
This is what I set in my nginx vhost:

```  
        ## FINAL
        location / {
                proxy_set_header X-Forwarded-Ssl on;

                proxy_pass http://127.0.0.1:5000;
                include /etc/nginx/proxy_params;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection $connection_upgrade;
                proxy_set_header X-Postal-Track-Host 1;
        }
```

**Comment by Bob-emblibrary** [comment]
Still an issue.
